src/item/enchantment/EnchantmentHelper.php

```diff
@@ -0,0 +1,221 @@
+					$block = $world->getBlock($tablePos->add($x, $y, $z));
+					if($block->getTypeId() !== BlockTypeIds::BOOKSHELF){
+						continue;
+					}
+
+					// Calculate the coordinates of the space between the bookshelf and the enchanting table
+					$spaceX = max(min($x, 1), -1);
+					$spaceZ = max(min($z, 1), -1);
+
+					$lowerSpaceBlock = $world->getBlock($tablePos->add($spaceX, 0, $spaceZ));
+					if($lowerSpaceBlock->getTypeId() !== BlockTypeIds::AIR){
+						break;
+					}
+					$upperSpaceBlock = $world->getBlock($tablePos->add($spaceX, 1, $spaceZ));
+					if($upperSpaceBlock->getTypeId() !== BlockTypeIds::AIR){
+						break;
+					}
+
+					$bookshelfCount++;
+					if($bookshelfCount === self::MAX_BOOKSHELF_COUNT){
+						return $bookshelfCount;
+					}
+				}
+			}
+		}
+
+		return $bookshelfCount;
+	}
+
+	private static function createEnchantOption(Random $random, Item $inputItem, int $optionCost, int $slot) : EnchantmentOption{
+		$cost = $optionCost;
+
+		$enchantability = self::getEnchantability($inputItem);
+		$cost = $cost + $random->nextRange(0, $enchantability >> 2) + $random->nextRange(0, $enchantability >> 2) + 1;
+		// Random bonus for enchantment cost between 0.85 and 1.15
+		$bonus = 1 + ($random->nextFloat() + $random->nextFloat() - 1) * 0.15;
+		$cost = (int) round($cost * $bonus);
+
+		$resultEnchantments = [];
+		$availableEnchantments = self::getAvailableEnchantments($cost, $inputItem);
+
+		$lastEnchantment = self::getRandomWeightedEnchantment($random, $availableEnchantments);
+		if ($lastEnchantment !== null) {
+			$resultEnchantments[] = $lastEnchantment;
+
+			// With probability (cost + 1) / 50, continue adding enchantments
+			while($random->nextFloat() <= ($cost + 1) / 50){
+				// Remove from the list of available enchantments anything that conflicts
+				// with previously-chosen enchantments
+				$availableEnchantments = array_filter(
+					$availableEnchantments,
+					function(EnchantmentInstance $e) use ($lastEnchantment){
+						return $e->getType() !== $lastEnchantment->getType() &&
+							$e->getType()->isCompatibleWith($lastEnchantment->getType());
+					}
+				);
+
+				$lastEnchantment = self::getRandomWeightedEnchantment($random, $availableEnchantments);
+				if($lastEnchantment === null){
+					break;
+				}
+
+				$resultEnchantments[] = $lastEnchantment;
+				$cost >>= 1;
+			}
+		}
+
+		return new EnchantmentOption($optionCost, $slot, $resultEnchantments, self::getRandomOptionName($random));
+	}
+
+	private static function getEnchantability(Item $item) : int{
+		if($item instanceof TieredTool){
+			return $item->getTier()->getEnchantability();
+		}
+		if($item instanceof Armor){
+			return $item->getMaterial()->getEnchantability();
+		}
+		return 1;
+	}
+
+	/**
+	 * @return EnchantmentInstance[]
+	 */
+	private static function getAvailableEnchantments(int $cost, Item $item) : array{
+		$list = [];
+
+		foreach(VanillaEnchantments::getAll() as $enchantment){
+			if($enchantment->isTreasure()){
```

Comment:
I don't think the use of `VanillaEnchantments` here makes sense, nor `isTreasure()`. Better if we have a dynamic register of enchantments that can be given by the enchanting table, which would also permit plugins to a) change which vanilla enchantments can be given, and b) allow custom enchantments to appear in the enchanting table.